packages/plugin-mongo/src/persistence.ts

```diff
@@ -0,0 +1,235 @@
+import type { Cursor } from "@apibara/protocol";
+import type { ClientSession, Db } from "mongodb";
+
+export type CheckpointSchema = {
+  id: string;
+  orderKey: number;
+  uniqueKey?: `0x${string}`;
+};
+
+export type FilterSchema = {
+  id: string;
+  filter: Record<string, unknown>;
+  fromBlock: number;
+  toBlock: number | null;
+};
+
+export const checkpointCollectionName = "checkpoints";
+export const filterCollectionName = "filters";
+
+export async function initializePersistentState(
+  db: Db,
+  session: ClientSession,
+) {
+  const checkpoint = await db.createCollection<CheckpointSchema>(
+    checkpointCollectionName,
+    { session },
+  );
+  const filter = await db.createCollection<FilterSchema>(filterCollectionName, {
+    session,
+  });
+
+  await checkpoint.createIndex({ id: 1 }, { session });
+  await filter.createIndex({ id: 1, fromBlock: 1 }, { session });
+}
+
+export async function persistState<TFilter>(props: {
+  db: Db;
+  session: ClientSession;
+  endCursor: Cursor;
+  filter?: TFilter;
+  indexerName?: string;
+}) {
+  const { db, session, endCursor, filter, indexerName = "default" } = props;
+
+  const mongoPersistence = new MongoPersistence<TFilter>(
+    db,
+    session,
+    indexerName,
+  );
+
+  await mongoPersistence.put({ cursor: endCursor, filter });
```

Comment:
Can we avoid creating the `MongoPersistence` object and just implement the method inline? Add helper functions to avoid duplicate code but only if needed.